Data check

syntrust opened the pull request:
This PR implements feature https://github.com/ethstorage/es-node/issues/384.

## Usage
Refer to https://github.com/ethstorage/es-node/tree/data-check/ethstorage/scanner#usage

## Tests

Test results on L1 with the default settings (check meta) shows 8192 kvs can be checked within about 350ms:
`--scanner.mode 1 --scanner.interval 3 --scanner.batch-size 8192`


```
INFO [04-08|10:51:51.077] Scanner update loop received event - all shards done.
INFO [04-08|10:51:51.077] Scanner started                          interval=3m0s
INFO [04-08|10:51:51.077] Scan batch started                       mode=1
INFO [04-08|10:51:51.103] Scanner: query KV entry count done       kvEntryCount=2,012,738
INFO [04-08|10:51:51.137] Scanner: query local KV entries done     localKVs="[0 1 2... 2012735 2012736 2012737]"
INFO [04-08|10:51:51.423] Scanner: query KV meta done              kvsInBatch="[0 1 2... 8189 8190 8191]"
INFO [04-08|10:51:51.428] Scanner: scan batch done                 from=0 to=8191 count=8192 nextKvIndex=8192
INFO [04-08|10:51:51.428] Scan batch done                          duration=351.132629ms
INFO [04-08|10:54:51.078] Scan batch started                       mode=1
INFO [04-08|10:54:51.104] Scanner: query KV entry count done       kvEntryCount=2,012,738
INFO [04-08|10:54:51.139] Scanner: query local KV entries done     localKVs="[0 1 2... 2012735 2012736 2012737]"
INFO [04-08|10:54:51.426] Scanner: query KV meta done              kvsInBatch="[8192 8193 8194... 16381 16382 16383]"
INFO [04-08|10:54:51.431] Scanner: scan batch done                 from=8192 to=16383 count=8192 nextKvIndex=16384
INFO [04-08|10:54:51.431] Scan batch done                          duration=352.769547ms
INFO [04-08|10:57:51.078] Scan batch started                       mode=1
INFO [04-08|10:57:51.103] Scanner: query KV entry count done       kvEntryCount=2,012,738
INFO [04-08|10:57:51.114] Scanner: query local KV entries done     localKVs="[0 1 2... 2012735 2012736 2012737]"
INFO [04-08|10:57:51.392] Scanner: query KV meta done              kvsInBatch="[16384 16385 16386... 24573 24574 24575]"
INFO [04-08|10:57:51.397] Scanner: scan batch done                 from=16384 to=24575 count=8192 nextKvIndex=24576
INFO [04-08|10:57:51.397] Scan batch done                          duration=319.084209ms
```

Test results on L2 show that all the data can be scanned with mode 2 (check blobs)  in 12 minutes and then started over again. (`--scanner.mode 2 --scanner.batch-size 4096 --scanner.interval 4`)
```
INFO [04-07|10:52:47.070] Scanner update loop received event - all shards done.
INFO [04-07|10:52:47.070] Scanner started                          interval=4m0s
INFO [04-07|10:52:47.070] Scan batch started
INFO [04-07|10:52:47.095] Scanner: query KV entry count done       kvEntryCount=15328
INFO [04-07|10:52:47.098] Scanner: query local KV entries done     localKVs="[0 1 2... 15325 15326 15327]"
INFO [04-07|10:52:47.379] Scanner: query KV meta done              kvsInBatch="[0 1 2... 4093 4094 4095]"
INFO [04-07|10:56:02.040] Scanner: scan batch done                 from=0 to=4095 count=4096 nextKvIndex=4096
INFO [04-07|10:56:02.040] Scan batch done                          duration=3m14.970046939s
INFO [04-07|10:56:47.071] Scan batch started
INFO [04-07|10:56:47.113] Scanner: query KV entry count done       kvEntryCount=15328
INFO [04-07|10:56:47.115] Scanner: query local KV entries done     localKVs="[0 1 2... 15325 15326 15327]"
INFO [04-07|10:56:47.392] Scanner: query KV meta done              kvsInBatch="[4096 4097 4098... 8189 8190 8191]"
INFO [04-07|10:59:56.598] Scanner: scan batch done                 from=4096 to=8191 count=4096 nextKvIndex=8192
INFO [04-07|10:59:56.598] Scan batch done                          duration=3m9.5276851s
INFO [04-07|11:00:47.071] Scan batch started
INFO [04-07|11:00:47.096] Scanner: query KV entry count done       kvEntryCount=15328
INFO [04-07|11:00:47.100] Scanner: query local KV entries done     localKVs="[0 1 2... 15325 15326 15327]"
INFO [04-07|11:00:47.383] Scanner: query KV meta done              kvsInBatch="[8192 8193 8194... 12285 12286 12287]"
INFO [04-07|11:04:04.884] Scanner: scan batch done                 from=8192 to=12287 count=4096 nextKvIndex=12288
INFO [04-07|11:04:04.884] Scan batch done                          duration=3m17.813780791s
INFO [04-07|11:04:47.071] Scan batch started
INFO [04-07|11:04:47.096] Scanner: query KV entry count done       kvEntryCount=15328
INFO [04-07|11:04:47.099] Scanner: query local KV entries done     localKVs="[0 1 2... 15325 15326 15327]"
INFO [04-07|11:04:47.305] Scanner: query KV meta done              kvsInBatch="[12288 12289 12290... 15325 15326 15327]"
INFO [04-07|11:07:18.437] Scanner: scan batch done                 from=12288 to=15327 count=3040 nextKvIndex=15328
INFO [04-07|11:07:18.437] Scan batch done                          duration=2m31.366055679s
INFO [04-07|11:08:47.070] Scan batch started
INFO [04-07|11:08:47.096] Scanner: query KV entry count done       kvEntryCount=15328
INFO [04-07|11:08:47.100] Scanner: query local KV entries done     localKVs="[0 1 2... 15325 15326 15327]"
INFO [04-07|11:08:47.100] Scanner: scan batch start over
INFO [04-07|11:08:47.399] Scanner: query KV meta done              kvsInBatch="[0 1 2... 4093 4094 4095]"
INFO [04-07|11:12:07.595] Scanner: scan batch done                 from=0     to=4095  count=4096 nextKvIndex=4096
INFO [04-07|11:12:07.595] Scan batch done                          duration=3m20.524446547s
```

Tests on a private L2 devnet demonstrated that a bad blob is detected and fixed immediately (mode 2):
```
INFO [04-08|17:46:51.122] Scanner: query KV entry count done       kvEntryCount=5
INFO [04-08|17:46:51.122] Scanner: query local KV entries done     localKVs="[0 1 2 3 4]"
INFO [04-08|17:46:51.368] Scanner: query KV meta done              kvsInBatch="[0 1 2 3 4]"
ERROR[04-08|17:46:51.599] Scanner: read blob error                 kvIndex=1 commit=0x01a1f264cca8a03e5d8713b6241e6b6c8e1ee95943f3156c0000000000000000 err="mismatched commit"
INFO [04-08|17:46:51.797] Download blob from RPC done              kvIndex=1 commit=0x01a1f264cca8a03e5d8713b6241e6b6c8e1ee95943f3156c0000000000000000
INFO [04-08|17:46:51.838] Fix blob successfully!                   kvIndex=1 commit=0x01a1f264cca8a03e5d8713b6241e6b6c8e1ee95943f3156c8000000000000000
INFO [04-08|17:46:51.977] Scanner: scan batch done                 from=0 to=4 count=5 nextKvIndex=5
INFO [04-08|17:46:51.977] Scan batch done                          duration=1.101783641s
INFO [04-08|17:47:50.314] P2P Summary                              activePeers=1 inbound=0 outbound=1
INFO [04-08|17:48:50.316] P2P Summary                              activePeers=1 inbound=0 outbound=1
INFO [04-08|17:49:50.316] P2P Summary                              activePeers=1 inbound=0 outbound=1
INFO [04-08|17:50:50.317] P2P Summary                              activePeers=1 inbound=0 outbound=1
INFO [04-08|17:50:50.879] Scan batch started                       mode=2
INFO [04-08|17:50:51.132] Scanner: query KV entry count done       kvEntryCount=5
INFO [04-08|17:50:51.132] Scanner: query local KV entries done     localKVs="[0 1 2 3 4]"
INFO [04-08|17:50:51.132] Scanner: scan batch start over
INFO [04-08|17:50:51.386] Scanner: query KV meta done              kvsInBatch="[0 1 2 3 4]"
INFO [04-08|17:50:51.606] Scanner: scan batch done                 from=0 to=4 count=5 nextKvIndex=5
INFO [04-08|17:50:51.606] Scan batch done                          duration=727.19179ms
```

Fix blob on mode 1:
```
INFO [04-09|10:15:00.983] Scanner started                          mode=1 interval=3m0s
INFO [04-09|10:15:00.983] Scan batch started
INFO [04-09|10:15:01.228] Scanner: query KV entry count done       kvEntryCount=5
INFO [04-09|10:15:01.228] Scanner: query local KV entries done     localKVs="[0 1 2 3 4]"
INFO [04-09|10:15:01.472] Scanner: query KV meta done              kvsInBatch="[0 1 2 3 4]"
ERROR[04-09|10:15:01.472] Scanner: read blob error                 kvIndex=2 commit=0x014a2b8ba99002af156690a91018c5dd7268e920cf6457880000000000000000 err="mismatched commit"
INFO [04-09|10:15:01.549] Download blob from RPC done              kvIndex=2 commit=0x014a2b8ba99002af156690a91018c5dd7268e920cf6457880000000000000000
INFO [04-09|10:15:01.589] Fix blob successfully!                   kvIndex=2 commit=0x014a2b8ba99002af156690a91018c5dd7268e920cf6457888000000000000000
INFO [04-09|10:15:01.589] Scanner: scan batch done                 from=0 to=4 count=5 nextKvIndex=5
INFO [04-09|10:15:01.589] Scan batch done                          duration=605.821458ms
INFO [04-09|10:16:00.370] P2P Summary                              activePeers=1 inbound=0 outbound=1
INFO [04-09|10:17:00.366] P2P Summary                              activePeers=1 inbound=0 outbound=1
INFO [04-09|10:18:00.363] P2P Summary                              activePeers=1 inbound=0 outbound=1
INFO [04-09|10:18:00.966] Scan batch started
INFO [04-09|10:18:01.200] Scanner: query KV entry count done       kvEntryCount=5
INFO [04-09|10:18:01.201] Scanner: query local KV entries done     localKVs="[0 1 2 3 4]"
INFO [04-09|10:18:01.201] Scanner: scan batch start over
INFO [04-09|10:18:01.434] Scanner: query KV meta done              kvsInBatch="[0 1 2 3 4]"
INFO [04-09|10:18:01.434] Scanner: scan batch done                 from=0 to=4 count=5 nextKvIndex=5
INFO [04-09|10:18:01.434] Scan batch done                          duration=468.413126ms
```

ping-ke commented on ethstorage/downloader/blob_disk_cache.go:
```diff
@@ -128,6 +128,7 @@ func (c *BlobDiskCache) GetKeyValueByIndex(idx uint64, hash common.Hash) []byte
 				data, err := c.store.Get(b.dataId)
 				if err != nil {
 					c.lg.Error("Failed to get kv from downloader cache", "kvIndex", idx, "id", b.dataId, "err", err)
+					c.lg.Error("Failed to get kv from downloader cache", "kvIndex", idx, "hash", b.hash, "expected", hash)
```
can we update the log instead of add a new line?

ping-ke commented on ethstorage/scanner/config.go:
```diff
@@ -0,0 +1,77 @@
+	flags := []cli.Flag{
+		cli.IntFlag{
+			Name:   ModeFlagName,
+			Usage:  "Data scan mode, 0: disabled, 1: check meta, 2: check blob",
```
can we enable both with different scanner.batch-size and scanner.interval?

syntrust replied:
Do you mean to check both meta and blob? It is unnecessary to check both because the source of the meta is the blob.

ping-ke commented on sync-l2.sh:
```diff
@@ -0,0 +1,10 @@
+# env ES_NODE_STORAGE_MINER=<miner> ./sync-l2.sh --kv_index <index>
+
+./sync.sh \
+  --storage.l1contract 0x64003adbdf3014f7E38FC6BE752EB047b95da89A \
```
remove this as it exit in another PR

qzhodl commented on ethstorage/scanner/scanner.go:
```diff
@@ -0,0 +1,133 @@
+				s.lg.Debug("syncEventCh closed, exiting update loop")
+				return
+			}
+			if syncDone.DoneType == protocol.AllShardDone {
```
What happens if a user chooses to sync and mine only a single shard while the network consists of multiple shards?

syntrust replied:
It only checks local shards that already synced

qzhodl commented on ethstorage/scanner/worker.go:
```diff
@@ -0,0 +1,167 @@
+		batchStart = 0
+		s.lg.Info("Scanner: scan batch start over")
+	}
+	batchEnd := batchStart + uint64(s.cfg.BatchSize)
```
Is there a scenario where the network has 4 shards, and if a user chooses to mine only shard 1 and shard 3, the kvIndex values become non-sequential?

syntrust replied:
Yes. Here is a discrete-shard case with local shards 0 and 2, kvEntryCount = 40, and 16 kv entries per shard.  Scanner config:
```
  --scanner.mode 1 \
  --scanner.interval 1 \
  --scanner.batch-size 20 \
```
>INFO [04-14|10:58:46.419] Scanner started                          mode=1 interval=1m0s batchSize=20
INFO [04-14|10:58:46.419] Scan batch started
INFO [04-14|10:58:46.713] Scanner: query KV entry count done       kvEntryCount=40
INFO [04-14|10:58:46.713] Scanner: local shards                    shards="[0 2]"
INFO [04-14|10:58:46.713] Scanner: query local KV entries done     localKVs="[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 32 33 34 35 36 37 38 39]"
INFO [04-14|10:58:47.023] Scanner: query KV meta done              kvsInBatch="[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 32 33 34 35]"
INFO [04-14|10:58:47.024] Scanner: scan batch done                 from=0 to=35 count=20 nextKvIndex=20
INFO [04-14|10:58:47.024] Scan batch done                          duration=605.464805ms
INFO [04-14|10:59:46.420] Scan batch started
INFO [04-14|10:59:47.079] Scanner: query KV entry count done       kvEntryCount=40
INFO [04-14|10:59:47.079] Scanner: local shards                    shards="[0 2]"
INFO [04-14|10:59:47.079] Scanner: query local KV entries done     localKVs="[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 32 33 34 35 36 37 38 39]"
INFO [04-14|10:59:47.380] Scanner: query KV meta done              kvsInBatch="[36 37 38 39]"
INFO [04-14|10:59:47.380] Scanner: scan batch done                 from=36 to=39 count=4  nextKvIndex=24
INFO [04-14|10:59:47.380] Scan batch done                          duration=960.576149ms
INFO [04-14|11:00:45.731] P2P Summary                              activePeers=1 inbound=0 outbound=1
INFO [04-14|11:00:46.419] Scan batch started
INFO [04-14|11:00:46.724] Scanner: query KV entry count done       kvEntryCount=40
INFO [04-14|11:00:46.724] Scanner: local shards                    shards="[0 2]"
INFO [04-14|11:00:46.724] Scanner: query local KV entries done     localKVs="[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 32 33 34 35 36 37 38 39]"
INFO [04-14|11:00:46.724] Scanner: scan batch start over
INFO [04-14|11:00:47.029] Scanner: query KV meta done              kvsInBatch="[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 32 33 34 35]"
INFO [04-14|11:00:47.029] Scanner: scan batch done                 from=0  to=35 count=20 nextKvIndex=20
INFO [04-14|11:00:47.029] Scan batch done                          duration=609.990212ms

qzhodl commented on ethstorage/scanner/worker.go:
```diff
@@ -0,0 +1,167 @@
+			s.lg.Error("Scanner: read blob error", "kvIndex", kvIndex, "commit", commit.Hex(), "err", err)
+			if err == es.ErrCommitMismatch {
+				if s.cfg.EsRpc == "" {
+					s.lg.Warn("Scanner: unable to fix blob: no RPC endpoint provided")
```
Does this error appear only once? Is there a way to consistently log this error whenever we fail to recover or fix the corrupted data?

syntrust replied:
Technically speaking, it appears in each scan interval. If this is not enough, should we make it a repeated summary like `Mining stats` or display it on the dashboard?

syntrust replied:
Print stats info per minute as in following log:
```
INFO [04-18|12:29:09.544] Scanner stats                            kvStored=5 mismatched=8 fixed=2 failed=6
ERROR[04-18|12:29:09.544] Scanner error happened                   kvIndex=1 error="failed to fix blob: failed to fetch blob: kvIndex=1, commit=01a1f264cca8a03e5d8713b6241e6b6c8e1ee95943f3156c0000000000000000, Post \"http://localhost:9545\": dial tcp [::1]:9545: connect: connection refused"
ERROR[04-18|12:29:09.544] Scanner error happened                   kvIndex=2 error="failed to fix blob: failed to fetch blob: kvIndex=2, commit=014a2b8ba99002af156690a91018c5dd7268e920cf6457880000000000000000, Post \"http://localhost:9545\": dial tcp [::1]:9545: connect: connection refused"
```

qzhodl commented on ethstorage/scanner/worker.go:
```diff
@@ -0,0 +1,167 @@
+					continue
+				}
+				// try to fix the blob
+				if err := s.fixKv(kvIndex, commit); err != nil {
```
Are there any potential race conditions when the downloader performs blob update actions?

syntrust replied:
No. We have lock in StorageManager.TryWrite

syntrust commented on ethstorage/scanner/worker.go:
```diff
@@ -0,0 +1,214 @@
+func (s *Worker) fixKv(kvIndex uint64, commit common.Hash) error {
+	s.lg.Info("Scanner: try to fix blob", "kvIndex", kvIndex, "commit", commit)
+	commitToFetchBlob := commit
+	newCommit, err := s.sm.CheckMeta(kvIndex)
```
The commit could have been updated since it was downloaded earlier.

qzhodl commented on ethstorage/scanner/worker.go:
```diff
@@ -0,0 +1,214 @@
+			kvIdx := shardId*kvEntries + k
+			if kvIdx < kvEntryCount {
+				// only check non-empty data
+				localKvs = append(localKvs, kvIdx)
```
This approach can lead to high memory usage when dealing with large shards. For instance, a 2 TB shard would consume around 122 MB of memory, which may become a bottleneck depending on the node’s resource constraints.

qzhodl commented on ethstorage/scanner/worker.go:
```diff
@@ -0,0 +1,214 @@
+			}
+		}
+	}
+	s.nextKvIndex = nextKvIndex
```
The variable name `nextKvIndex` is misleading since it is actually not the next kv index but the index of next kv index

syntrust replied:
fixed

qzhodl commented on ethstorage/scanner/worker.go:
```diff
@@ -0,0 +1,214 @@
+		// in case the commit in the contract is same with the one in the storage, and both updated lately
+		commitToFetchBlob = newCommit
+	}
+	blob, err := s.fetchBlob(kvIndex, commitToFetchBlob)
```
Can we move this part into TryWriteWithMetaCheck?

syntrust replied:
fixed

syntrust commented:
Recent main changes:

1. Determine kv entries in a batch without loading all the indices in memory.
2. Check kv entries only when the block that it updated is finalized.
3. Upgrade CommitMismatchError to provide details.
4. Improve scanner statistics so that it is more informative.

syntrust commented:
Log updated:
```
INFO [05-16|19:17:19.109] Storage node started
INFO [05-16|19:17:19.109] Scanner update loop received event - all shards done.
INFO [05-16|19:17:19.109] Scanner started                          mode=2 interval=1m0s batchSize=10
INFO [05-16|19:17:19.109] Sync done
INFO [05-16|19:17:19.109] Storage sync done                        subTaskCount=3
INFO [05-16|19:17:19.109] Sync done
INFO [05-16|19:17:19.109] Storage sync done                        subTaskCount=3
INFO [05-16|19:17:19.109] Scan batch started
INFO [05-16|19:17:19.109] Scanner: local storage info              lastKvIdx=42 shards="[0 1 2]" kvEntriesPerShard=16
INFO [05-16|19:17:19.109] Storage sync in progress                 shardId=0 subTaskRemain=0 peerCount=0 progress=100.00% blobsSynced=16 blobsToSync=0 timeUsed=1h39m14s etaTimeLeft=0s
INFO [05-16|19:17:19.109] Scanner: batch index range               batchStart=0 batchEnd=10 kvsInBatch=[0-9]
INFO [05-16|19:17:19.109] Storage sync in progress                 shardId=1 subTaskRemain=0 peerCount=0 progress=100.00% blobsSynced=17 blobsToSync=0 timeUsed=1h39m14s etaTimeLeft=0s
INFO [05-16|19:17:19.109] Storage sync in progress                 shardId=2 subTaskRemain=0 peerCount=0 progress=100.00% blobsSynced=11 blobsToSync=0 timeUsed=1h38m23s etaTimeLeft=0s
INFO [05-16|19:17:19.109] Storage fill empty in progress           shardId=2 subTaskRemain=0 progress=100.00% emptyFilled=5 emptyToFill=0 timeUsed=0s       etaTimeLeft=0s
WARN [05-16|19:17:19.598] Scanner: commit mismatch found           kvIndex=2 error="commit mismatch: expected=019f4874db6d971ec1ed8b04065ec6fd6ad5287c6504c907, found=01cd6acae63d9ec4ec931e94ea0500aa3d19c94a5a1ea027"
INFO [05-16|19:17:19.598] Scanner: try to fix blob                 kvIndex=2 commit=019f48..000000
ERROR[05-16|19:17:20.051] Scanner: fix blob error                  kvIndex=2 error="failed to write KV: kvIndex=2, commit=019f4874db6d971ec1ed8b04065ec6fd6ad5287c6504c9070000000000000000, fetch blob error: kvIdx=2, commit=019f4874db6d971ec1ed8b04065ec6fd6ad5287c6504c9070000000000000000, no peer can be used to send requests"
INFO [05-16|19:17:20.581] Scanner: scan batch done                 scanned=[0-9] count=10 nextIndexOfKvIdx=10
INFO [05-16|19:17:20.581] Scan batch done                          duration=1.47263853s
INFO [05-16|19:17:23.612] Connected to peer                        peer=16Uiu2HAmBHDJeHr26iisJsDzSc6qUQFzYfMyv2er5iD1kZ4AhgKF Direction=Outbound addr=/ip4/127.0.0.1/tcp/9222
INFO [05-16|19:17:23.653] Connected to peer                        peer=16Uiu2HAmBHDJeHr26iisJsDzSc6qUQFzYfMyv2er5iD1kZ4AhgKF Direction=Outbound addr=/ip4/10.37.38.183/tcp/9222
INFO [05-16|19:18:18.443] P2P Summary                              activePeers=1 inbound=0 outbound=1
INFO [05-16|19:18:19.110] Scanner stats                            localKvs=shard0[0-15],shard1[16-31],shard2[32-42] localKvsCount=43 mismatched=[2] fixed=[] failed=[2]
ERROR[05-16|19:18:19.110] Scanner error happened earlier           kvIndex=2 error="failed to fix blob: failed to write KV: kvIndex=2, commit=019f4874db6d971ec1ed8b04065ec6fd6ad5287c6504c9070000000000000000, fetch blob error: kvIdx=2, commit=019f4874db6d971ec1ed8b04065ec6fd6ad5287c6504c9070000000000000000, no peer can be used to send requests"
INFO [05-16|19:18:19.110] Scan batch started
INFO [05-16|19:18:19.110] Scanner: local storage info              lastKvIdx=42 shards="[0 1 2]" kvEntriesPerShard=16
INFO [05-16|19:18:19.110] Scanner: batch index range               batchStart=10 batchEnd=20 kvsInBatch=[10-19]
INFO [05-16|19:18:20.010] Scanner: scan batch done                 scanned=[10-19] count=10 nextIndexOfKvIdx=20
INFO [05-16|19:18:20.010] Scan batch done                          duration=899.888109ms
INFO [05-16|19:19:18.444] P2P Summary                              activePeers=1 inbound=0 outbound=1
INFO [05-16|19:19:19.111] Scan batch started
INFO [05-16|19:19:19.111] Scanner: local storage info              lastKvIdx=42 shards="[0 1 2]" kvEntriesPerShard=16
INFO [05-16|19:19:19.111] Scanner: batch index range               batchStart=20 batchEnd=30 kvsInBatch=[20-29]
INFO [05-16|19:19:19.111] Scanner stats                            localKvs=shard0[0-15],shard1[16-31],shard2[32-42] localKvsCount=43 mismatched=[2] fixed=[] failed=[2]
ERROR[05-16|19:19:19.111] Scanner error happened earlier           kvIndex=2 error="failed to fix blob: failed to write KV: kvIndex=2, commit=019f4874db6d971ec1ed8b04065ec6fd6ad5287c6504c9070000000000000000, fetch blob error: kvIdx=2, commit=019f4874db6d971ec1ed8b04065ec6fd6ad5287c6504c9070000000000000000, no peer can be used to send requests"
WARN [05-16|19:19:19.524] Scanner: commit mismatch found           kvIndex=22 error="commit mismatch: expected=01fa4919ea8bf1fd99d5ccdafef75fdbf17fcb56a8fefe7b, found=017ea06714a6aedb07ba7013d4336d4e9499f58b1559f5da"
INFO [05-16|19:19:19.524] Scanner: try to fix blob                 kvIndex=22 commit=01fa49..000000
INFO [05-16|19:19:19.825] Scanner: fixed blob successfully!        kvIndex=22
INFO [05-16|19:19:20.215] Scanner: scan batch done                 scanned=[20-29] count=10 nextIndexOfKvIdx=30
INFO [05-16|19:19:20.215] Scan batch done                          duration=1.103653618s
INFO [05-16|19:20:18.444] P2P Summary                              activePeers=1 inbound=0 outbound=1
INFO [05-16|19:20:19.112] Scanner stats                            localKvs=shard0[0-15],shard1[16-31],shard2[32-42] localKvsCount=43 mismatched=[2,22] fixed=[22] failed=[2]
ERROR[05-16|19:20:19.112] Scanner error happened earlier           kvIndex=2  error="failed to fix blob: failed to write KV: kvIndex=2, commit=019f4874db6d971ec1ed8b04065ec6fd6ad5287c6504c9070000000000000000, fetch blob error: kvIdx=2, commit=019f4874db6d971ec1ed8b04065ec6fd6ad5287c6504c9070000000000000000, no peer can be used to send requests"
INFO [05-16|19:20:19.112] Scan batch started
INFO [05-16|19:20:19.112] Scanner: local storage info              lastKvIdx=42 shards="[0 1 2]" kvEntriesPerShard=16
INFO [05-16|19:20:19.112] Scanner: batch index range               batchStart=30 batchEnd=40 kvsInBatch=[30-39]
INFO [05-16|19:20:19.934] Scanner: scan batch done                 scanned=[30-39] count=10 nextIndexOfKvIdx=40
INFO [05-16|19:20:19.934] Scan batch done                          duration=821.89997ms
INFO [05-16|19:21:18.444] P2P Summary                              activePeers=1 inbound=0 outbound=1
INFO [05-16|19:21:19.112] Scanner stats                            localKvs=shard0[0-15],shard1[16-31],shard2[32-42] localKvsCount=43 mismatched=[2,22] fixed=[22] failed=[2]
ERROR[05-16|19:21:19.112] Scanner error happened earlier           kvIndex=2  error="failed to fix blob: failed to write KV: kvIndex=2, commit=019f4874db6d971ec1ed8b04065ec6fd6ad5287c6504c9070000000000000000, fetch blob error: kvIdx=2, commit=019f4874db6d971ec1ed8b04065ec6fd6ad5287c6504c9070000000000000000, no peer can be used to send requests"
INFO [05-16|19:21:19.112] Scan batch started
INFO [05-16|19:21:19.112] Scanner: local storage info              lastKvIdx=42 shards="[0 1 2]" kvEntriesPerShard=16
INFO [05-16|19:21:19.112] Scanner: batch index range               batchStart=40 batchEnd=43 kvsInBatch=[40-42]
INFO [05-16|19:21:19.742] Scanner: scan batch done                 scanned=[40-42] count=3  nextIndexOfKvIdx=43
INFO [05-16|19:21:19.742] Scan batch done                          duration=630.184694ms
INFO [05-16|19:22:18.446] P2P Summary                              activePeers=1 inbound=0 outbound=1
INFO [05-16|19:22:19.113] Scan batch started
INFO [05-16|19:22:19.113] Scanner: local storage info              lastKvIdx=42 shards="[0 1 2]" kvEntriesPerShard=16
INFO [05-16|19:22:19.113] Scanner: scan batch start over
INFO [05-16|19:22:19.113] Scanner: batch index range               batchStart=0  batchEnd=10 kvsInBatch=[0-9]
INFO [05-16|19:22:19.113] Scanner stats                            localKvs=shard0[0-15],shard1[16-31],shard2[32-42] localKvsCount=43 mismatched=[2,22] fixed=[22] failed=[2]
ERROR[05-16|19:22:19.113] Scanner error happened earlier           kvIndex=2  error="failed to fix blob: failed to write KV: kvIndex=2, commit=019f4874db6d971ec1ed8b04065ec6fd6ad5287c6504c9070000000000000000, fetch blob error: kvIdx=2, commit=019f4874db6d971ec1ed8b04065ec6fd6ad5287c6504c9070000000000000000, no peer can be used to send requests"
WARN [05-16|19:22:19.549] Scanner: commit mismatch found           kvIndex=2  error="commit mismatch: expected=019f4874db6d971ec1ed8b04065ec6fd6ad5287c6504c907, found=01cd6acae63d9ec4ec931e94ea0500aa3d19c94a5a1ea027"
INFO [05-16|19:22:19.549] Scanner: try to fix blob                 kvIndex=2  commit=019f48..000000
INFO [05-16|19:22:19.857] Scanner: fixed blob successfully!        kvIndex=2
INFO [05-16|19:22:20.292] Scanner: scan batch done                 scanned=[0-9]   count=10 nextIndexOfKvIdx=10
INFO [05-16|19:22:20.292] Scan batch done                          duration=1.178881727s
```

qzhodl commented on ethstorage/scanner/worker.go:
```diff
@@ -0,0 +1,206 @@
+
+		if !found {
+			// The shard does not exist locally
+			s.lg.Warn("Scanner: blob not found", "kvIndex", kvIndex, "commit", commit)
```
The case should not happen, and an error should be print here

qzhodl commented on ethstorage/downloader/downloader.go:
```diff
@@ -259,14 +260,14 @@ func (s *Downloader) download() {
-		return
-	}
+	s.log.Warn("Download blobs", "trackHead", trackHead, "lastDownloadBlock", s.lastDownloadBlock)
+	// if (s.lastDownloadBlock > 0) && (trackHead-s.lastDownloadBlock > int64(s.minDurationForBlobsRequest)) {
```
This should not be commented

qzhodl reviewed (approved):
LGTM

syntrust commented on ethstorage/scanner/utils.go:
```diff
@@ -0,0 +1,128 @@
+	for _, shard := range shards {
+		if shard*kvEntries > lastKvIdx {
+			// skip empty shards
+			break
```
When lastKvIdx is crossed, we should end the loop

syntrust commented on ethstorage/scanner/worker.go:
```diff
@@ -0,0 +1,203 @@
+		lg.Info("Scanner: scan batch start over")
+	}
+	batchEnd := batchStart + batchSize
+	if batchEnd > totalEntries {
```
We prefer not to restart from kv 0 because during the bootstrap stage, there may be fewer kvs compared to a scan batch, which could result in unnecessary repeated checks on the data.

syntrust commented on ethstorage/scanner/worker.go:
```diff
@@ -0,0 +1,203 @@
+	var kvsInBatch []uint64
+	var currentIndex uint64
+out:
+	for _, shardId := range shards {
```
Fixed. In commit 5468088 the boundaries have been computed beforehand to avoid unnecessary checks in loops.
